doc/neps/nep-0050-scalar-promotion.rst

```diff
@@ -0,0 +1,729 @@
+  # and:
+  np.add(np.arange(10, dtype=np.uint8), np.int64(1))
+
+Will return an int64 array because the type of ``np.int64(1)`` is strictly
+honoured.
+
+
+Impact on operators involving Python ``int``, ``float``, and ``complex``
+------------------------------------------------------------------------
+
+This NEP attempts to preserve the convenience that the old behaviour
+gave when working with literal values.
+The current value-based logic had some nice properties when "untyped",
+literal Python scalars are involved::
+
+  np.arange(10, dtype=np.int8) + 1  # returns an int8 array
+  np.array([1., 2.], dtype=np.float32) * 3.5  # returns a float32 array
+
+But led to complexity when it came to "unrepresentable" values::
+
+  np.arange(10, dtype=np.int8) + 256  # returns int16
+  np.array([1., 2.], dtype=np.float32) * 1e200  # returns float64
+
+The proposal is to preserve this behaviour for the most part.  This is achieved
+by considering Python ``int``, ``float``, and ``complex`` to be "weakly" typed
+in these operations.
+However, to mitigate user surprises, we plan to make conversion to the new type
+more strict:  This means that the results will be unchanged in the first
+two examples.  For the second one, the results will be the following::
+
+  np.arange(10, dtype=np.int8) + 256  # raises a TypeError
+  np.array([1., 2.], dtype=np.float32) * 1e200  # warning and returns infinity
+
+The second one will warn because ``np.float32(1e200)`` overflows to infinity.
+It will then do the calculation with ``inf`` as normally.
+
+
+.. admonition:: Behaviour in other libraries
+
+   Overflowing in the conversion rather than raising an error is a choice;
+   it is one that is the default in most C setups (similar to NumPy C can be
+   set up to raise an error due to the overflow, however).
+   It is also for example the behaviour of ``pytorch`` 1.10.
+
+
+
+Backward compatibility
+======================
+
+In general, code which only uses the default dtypes float64, or int32/int64
+or more precise ones should not be affected.
+
+However, the proposed changes will modify results in quite a few cases where
+0-D or scalar values (with non-default dtypes) are mixed.
+In many cases, these will be bug-fixes, however, there are certain changes
+which may be particularly interesting.
+
+The most important failure is probably the following example::
+
+  arr = np.arange(100, dtype=np.uint8)  # storage array with low precision
+  value = arr[10]
+
+  # calculation continues with "value" without considering where it came from
+  value * 100
+
+Where previously the ``value * 100`` would cause an up-cast to int32/int64
+(because value is a scalar).  The new behaviour will preserve the lower
+precision unless explicitly dealt with (just as if ``value`` was an array).
+This can lead to integer overflows and thus incorrect results beyond precision.
+In many cases this may be silent, although NumPy usually gives warnings for the
+scalar operators.
```

Comment:
This sound like a really bad idea to me. I'm sorry to be so negative on this, but this could well lead to silent and catastrophic failures & data corruption, and IMO poses a substantial risk of blemishing numpy's reputation if not handled with extreme care.

Reply:
Note that I am planning to add at least an *optional* warning tthat will notify of places where reduced precision occurs (I am not sure it would be 100%, but it should be good enough in practice).

Reply:
Why will these warnings be optional?  This is the case I am by far the most worried about.

Reply:
Note that there are two warnings involved here:
* Overflow warnings are normally given for NumPy scalar operations (which are the main affected ones here).  (They are not given for array operations.)
* Warnings about the actual change, which I was thinking should probably be opt-in.

For the second, I was aiming for opt-in currently, because I think they will be very noisy and cumbersome to avoid.  E.g. for `float32 >= 2.1` where the behaviour might change, but practically never does.  Should we do the operation twice to check? Give a warning when it probably doesn't matter?

Maybe we could check whether always warnings should happen for more particularly surprising changes, right now I hoped the scalar integer overflow warnings cover the worst of it though.